killproc.8
.\"
.\" Copyright 1994-2000 Werner Fink, 1996-2000 SuSE GmbH Nuernberg, Germany.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.TH KILLPROC 8 "Nov 10, 2000" "Version 1.16" "The SuSE boot concept"
.UC 8
.SH NAME
Killproc \- Send signals to processes by full path name
.SH SYNOPSIS
.\"
.HP
.B killproc
.RB [ \-vqLN ]
.RB [ \-g | -G ]
.RB [ \-x ]
.RB [ \-p\ \fIpid_file\fP ]
.RB [ \-i\ \fIingnore_file\fP ]
.RB [ \-c\ \fIroot\fP ]
.RB [ \-t\ \fI<sec>\fP ]
.RB [ \-\fI<SIG>\fP ]
.I /full/path/to/executable
.HP
.B killproc
.B \-n
.RB [ \-vq ]
.RB [ \-g | -G ]
.RB [ \-t\ \fI<sec>\fP ]
.RB [ \-\fI<SIG>\fP ]
.I name_of_kernel_thread
.HP
.B killproc
.B \ \ 
.RB [ \-vq ]
.RB [ \-g | -G ]
.RB [ \-t\ \fI<sec>\fP ]
.RB [ \-\fI<SIG>\fP ]
.I basename_of_executable
.HP
.B killproc
.B \-l
.SH SYNOPSIS LSB 3\.1
.HP
.B killproc
.RB [ \-p\ \fIpid_file\fP ]
.I /full/path/to/executable
.RB [ \-\fI<SIG>\fP ]
.\"
.SH DESCRIPTION
.B killproc
sends signals to all processes that use the specified executable.
If no signal name is specified, the signal
.B SIGTERM
is sent. If this program is not called with the name
.B killproc
then
.B SIGHUP
is used. Note that if no signal is specified on the command line and the program
.B killproc
can not terminate a process with the default
.B SIGTERM
the signal
.B SIGKILL
is send after a few seconds (default is 5 seconds, see option
.BR \-t ).
If a program has been terminated successfully and a
.B verified
pid file was found, this pid file will be removed if the terminated
process didn't already do so.
.PP
.B killproc
does not use the pid to send a signal to a process but the full
path of the corresponding program which is used to identify the executable
.RB (see " proc" (5)).
Only if inode number
.RB (/proc/ <pid> /exe)
and full path name of the executable
.RB (/proc/ <pid> /cmdline)
are unavailable or if the executable has changed its zeroth argument,
.B killproc
uses the base name
.RB (/proc/ <pid> /stat)
to identify the running program. Note that if the option
.B \-n
for kernel thread is given, only
.RB (/proc/ <pid> /stat)
is used.  For this case an existing symbolic link
.RB (/proc/ <pid> /exe)
indicates that the
.B <pid>
is not a kernel thread.
.PP
The last variant of the
.B killproc
call is only with the basename
of the executable.  This is required by the Linux Standard Base Specification (LSB).
Note that this disables the comparision of the inodes of the
executable and the information found within the proc table
.RB (see " proc" (5)).
This implies that if no pid file is found it may happens that
programs or script with the same base name will killed instead.
.B killproc
tries to determine the pid of its parent and pid of the parent of its parent
to not to kill those two processes.
.PP
Extended functionality is provided by the
.B \-p
.IR pid_file
option (former option
.B \-f
changed due to the LSB specification).
If this option is specified,
.B killproc
tries to send the signal to the pid read from this file instead
of from the default pid file
.RB (/var/run/ <basename> .pid).
If the
.IR pid_file
does not exist,
.B killproc
assumes that the daemon is not running.
It is possible to use a process identity number instead
of a pid file.
.PP
For the possibility of having two different sessions of one binary
program, the option
.B \-i
.IR ignore_file
allows to specify a pid file which pid number is used to
.B ignore
all processes of corresponding process
.BR session .
.PP
The option
.B \-v
turns on verbosity about which signal is being sent.
The option
.B \-q
is
.B ignored
for compatibility with older versions.
If the option
.B \-G
is used, the signal will not only sent to the running executable,
but also to all members of the session that the specified executable
leads. Whereas the option
.B \-g
the signal sends to the group including the executable.
If unsure use
.BR \-G .
.PP
.\"
.SH REQUIRED
.TP
.IR /full/path/to/executable " or " name_of_kernel_thread
Specifies the executable to which the signal should be sent,
or alternatively, if the option
.B \-n
is used, the name of the kernel thread. This argument is
always required.
.\"
.SH OPTIONS
.TP
.B \-G
Sends the signal to all session followers (children) of the identified
process.
.TP
.B \-g
Sends the signal to all members of the session including the
identified process.
Note that usually the option
.B \-G
should be used.
.TP
.B \-L
This option causes symlinks to be followed, as the
like-named option in
.BR ls (1).
.BR Note : 
for the file name the original name of the program is used
instead of the name of the symbolic link.
.TP
.BI \-p " pid_file"
(Former option
.B \-f
changed due to the LSB specification.)
Use an alternate pid file instead of the default
.RB (/var/run/ <basename> .pid).
If the
.IR pid_file
does not exist,
.B killproc
assumes that the daemon is not running.
It is possible to use a process identity number instead
of a pid file.
.TP
.BI \-i " ignore_file"
The pid found in this file is used as
.B session
id of the same binary program which should be ignored by
.BR killproc .
.TP
.BI \-c " root"
Change root directory to
.I root
for services which have been started with this option by
.BR startproc (8).
.TP
.B \-n
This option indicates that a kernel thread should be
signaled.  In this case not the executable with its
full path name is required but the name of the kernel
thread.
.TP
.B \-N
With this option
the location of the executable is checked about NFS
file system and if true the
.BR stat (2)
system call is not applied on the
.B exe
symbolic link under
.BR /proc (5).
Otherwise
.B killproc
could be locked if the corresponding NFS server
is currently not online or available.
This implies that the inode number check between the exectuable
on the command line and the exectuable of the
.B exec
symbolic link will be skipped.
.\"
.TP
.B \-\fI<SIG>\fP
Signals  can be specified either by name
.RB (e.g. " -HUP" , " -SIGHUP" )
or by number
.RB (e.g. " -1" ).
.TP
.B \-t\fI<sec>\fP
The number
.I <sec>
specifies the seconds to wait between the default signal
.B SIGTERM
and the subsequentially signal
.B SIGKILL
if the first
.B SIGTERM
does not show any result within the
first few milli seconds.  This timeout defaults to
.B 5
seconds.  On the other hand if the signal
.B SIGTERM
was  explicitly used on the command line, the signal
.B SIGKILL
will be omitted, even if the process was not terminated
after the timeout has expired.
.TP
.B \-q
This option is ignored.
.TP
.B \-v
Be more verbose.
.\"
.TP
.B \-l
This option list all available signals and some of their
synonyms by their number and signal names to standard out.
and exits.
.TP
.B \-x
Tells
.B killproc
to use the upto 15 byte long filename name found in
.I /proc/<pid>/stat
to be able to identify scripts overwriting their command line.
.\"
.SH EXAMPLES
.TP
.B killproc -TERM /usr/sbin/sendmail
.IP
sends the signal
.B SIGTERM
to the running sendmail process. If a
pid file
.I sendmail.pid
exists in
.I /var/run/
then the pid from this file is being used after verification to
terminate the sendmail process.  Other running processes are
ignored.
.TP
.B killproc -p /var/myrun/lpd.pid /usr/sbin/lpd
.IP
sends the signal
.B SIGTERM
to the pid found in
.I /var/myrun/lpd.pid
if and only if this pid belongs to
.IR /usr/sbin/lpd .
If the named
.I /var/myrun/lpd.pid
does not exist,
.B killproc
assumes that the daemon of
.B /usr/sbin/lpd
is not running.  The exit status
is set to 0 for successfully delivering the default signals
.BR SIGTERM " and " SIGKILL
otherwise to 7 if the program was not running.  It is also
successful if
.B no
signal was specified and
.B no program
was there for Termination because it is already terminated.
.\"
.SH EXIT CODES
The exit codes have the following
.B LSB
conform conditions:
.PP
.RS 7
.IP 0 5
Success or program was not running (no signal specified)
.IP 1 5
Generic or unspecified error
.IP 2 5
Invalid or excess argument(s)
.IP 4 5
Insufficient privilege(s)
.IP 5 5
Program is not installed
.IP 7 5
Program was not running to receive the specified signal
.RE
.PP
In some error cases, diagnostic output is sent to standard error, or,
if standard error is not available,
.BR syslogd "(8) is being used."
.SH NOTE
.B killproc
is a replacment for the Bourne shell function
.B killproc
found in the widely used SysVinit package of
Miquel van Smoorenburg, <miquels@cistron.nl>.
.\"
.SH BUGS
Identifying a process based on the executable file and the
corresponding inode number only works if the process stays alive
during killproc's execution. Impure executables like shell scripts
(the inode number of the shell is not identical to that of the script)
and programs rewriting their zeroth argument may not be identified by
a file name.
.PP
.B Killproc
isn't able to signal processes being in the zombie state. Zombies are
processes which  arn't alive but listed in the process table to have
the exit status ready for the corresponding parent processes.
.\"
.SH FILES
.TP
.I /proc/
path to the proc file system
.RB (see " proc" (5)).
.TP
.I /etc/init.d/
path to the SuSE boot concept script base directory
as required by the Linux Standard  Base  Specification (LSB)
.RB (see " init.d" (7)).
.\"
.SH SEE ALSO
.BR startproc (8),
.BR checkproc (8),
.BR insserv (8),
.BR init.d (7),
.BR kill (1),
.BR skill (1),
.BR killall (8),
.BR killall5 (8),
.BR signal (7),
.BR proc (5).
.SH COPYRIGHT
1994-2005 Werner Fink,
1996-2005 SuSE GmbH Nuernberg, Germany.
.SH AUTHOR
Werner Fink <werner@suse.de>


killproc.c
/*
 * killproc.c   Kill all running processes of a named program.
 *
 * Usage:       killproc [-v] [-t<sec>] [-g|-G] [-SIG] /full/path/to/program
 *
 * Copyright 1994-2000 Werner Fink, 1996-2000 SuSE GmbH Nuernberg, Germany.
 * Copyright 2005 Werner Fink, 2005 SUSE LINUX Products GmbH, Germany.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Author:      Werner Fink <werner@suse.de>
 *
 * 1998/05/06 Florian La Roche: added "-g" option to kill process groups
 * 1998/05/06 Werner Fink: rework, added "-p" for pid files
 * 1998/15/09 Werner Fink: exit status for killing not running processes is 0
 * 1998/29/09 Werner Fink: Add kernel thread handling.
 * 2000/11/10 Werner Fink: LSB specs, logging
 * 2007/11/29 Werner Fink: ignore more than one pid
 */

#include "libinit.h"
#include "statx.h"

#define DEFSIG		"TERM"
#define OTHERSIG	"HUP"

#define USAGE		"Usage:\n"\
			"    %s [-v] [-q] [-L] [-g|-G] [-N] [-p pid_file] [-i ingnore_file] \\\n"\
			"        [-c root] [-t<sec>] [-SIG] /full/path/to/executable\n"\
		 	"    %s -l\n", we_are, we_are

static int do_kill(const char *name, const pid_t proc, const int sig,
		   const int group_leader, const int process_group);

static int quiet = 1, num;

int main(int argc, char **argv)
{
    int c, snum;
    struct stat st;
    list_t *list;
    char *fullname = NULL, *basename = NULL;
    char *pid_file = NULL, *ignore_file = NULL;
    char *root = NULL;
    extension char *iargv[argc];
    char *posixa, *posixb;	/* Don't fool me with posix correct */
    int process_group = 0, group_leader = 0, wait = 5, iargc = 0;
    unsigned short flags = (KILL|PIDOF|KSTOP);
    boolean pid_forced = false;
    boolean sig_forced = true;

    we_are = base_name(argv[0]);
    openlog (we_are, LOG_OPTIONS, LOG_FACILITY);
    for (c = 0; c < argc; c++)
	iargv[c] = (char*)0;

    /* If we are not called as killproc use HUP (e.g. for sigproc) */
    if (strcmp(we_are,"killproc") != 0)
	snum = signame_to_signum(OTHERSIG);
    else
	snum = signame_to_signum(DEFSIG);

    /*
     *  We should stat() fullname, because only the path identifies the executable.
     *  If there is one hardlink we have only to stat() the orignal executable.
     *  If there is more than one hardlink and we have to distinguish the
     *  executables by their swapname.  Note if the cmdline of some executables
     *  will changed by the running process its self the name is not clearly
     *  defined ... see libinit.c for more information.
     */

    c = argc;
    while (--c) {
	if (*(argv[c]) == '-') {
	    char *sig = argv[c];
	    int tmp, len = strlen(sig);
	    sig++;
	    if ( (tmp = atoi(sig)) > 0 && tmp < NSIG ) {
		memset(sig, 0, len);
		*sig = 'q';		/* set dummy option -q */
		snum = tmp;
		sig_forced = false;
		break;
	    } else if ( (tmp = signame_to_signum(sig)) > 0 ) {
		memset(sig, 0, len);
		*sig = 'q';		/* set dummy option -q */
		snum = tmp;
		sig_forced = false;
		break;
	    }
	}
    }

    posixa = getenv("_POSIX_OPTION_ORDER"); unsetenv("_POSIX_OPTION_ORDER");
    posixb = getenv("POSIXLY_CORRECT");     unsetenv("POSIXLY_CORRECT");
    opterr = 0;
    while ((c = getopt(argc, argv, "c:p:gGnNhlvqt:Li:x")) != -1) {
	switch (c) {
	    case 'c':
		if (optarg && optarg[0] != '-' && !root) {
		    root = optarg;
		} else
		    error(LSB_WRGSYN,"Option -c requires special root directory\n");
		break;
	    case 't':
		wait = atoi(optarg);
		if (wait < 1)
		    error(LSB_WRGSYN, USAGE);
		break;
	    case 'q':
		flags &= ~KSTOP;
		/* A signal which has been handled or the old but unused -q option */
		break;
	    case 'v':
		quiet = 0;
		break;
	    case 'L':
		flags |= FLWLINK;
		break;
	    case 'g':
		if (process_group)
		    error(LSB_WRGSYN, USAGE);
		group_leader++;
		break;
	    case 'G':
		if (group_leader)
		    error(LSB_WRGSYN, USAGE);
		process_group++;
		break;
	    case 'n':
		flags |= KTHREAD;
		break;
	    case 'N':
		init_nfs();
		break;
	    case 'x':
		flags |= STSCRPT;
		break;
	    case 'p':		/* Changed from -f to -p to fit startproc and LSB */
		/* Allocate here: address optarg (current *argv) isn't freeable */
		if (optarg && !pid_file) {
		    pid_file = xstrdup(optarg);
		} else
		    error(LSB_WRGSYN,"Option -p requires pid file to read pid from\n");
		break;
	    case 'l':
		list_signames();
		exit(0);
	    case 'i':
		/* Remember: address optarg (current *argv) */
		if (optarg && optarg[0] != '-') {
		    iargv[iargc++] = optarg;
		} else
		    error(LSB_WRGSYN,"Option -i requires pid file to read pid from\n");
		break;
	    case '?':
		error(LSB_WRGSYN, USAGE);
		break;
	    case 'h':
		error(0, USAGE);
		break;
	    default:
		break;
	}
    }
    if (posixa) setenv("_POSIX_OPTION_ORDER", posixa, 0);
    if (posixb) setenv("POSIXLY_CORRECT",     posixb, 0);

    argv += optind;
    argc -= optind;

    if (!*argv)
	error(LSB_WRGSYN, USAGE);

    if (root) {
	fullname = (char*) xmalloc(strlen(*argv)+strlen(root)+1);
	fullname = strcat(strcpy(fullname,root),*argv);
    } else
	fullname = *argv;

    if (flags & FLWLINK) {
	/* rlstat replaces the current fullname with that of the real file */
	if (rlstat(&fullname, &st, flags) < 0)
	    warn("cannot stat %s: %s\n", fullname, strerror(errno));
    }
    basename = base_name(fullname);

    if (*fullname != '/')
	flags |= KSHORT;
    else if (check4nfs(fullname))
	flags |= (KSHORT|KBASE);
    clear_nfs();

    if (!pid_file) {            /* the default pid file */
	if (root) {
	    pid_file = (char*) xmalloc(DEFPIDLEN+strlen(basename)+strlen(root)+1);
	    pid_file = strcpy(pid_file,root);
	    pid_file = strcat(strcat(strcat(pid_file,DEFPIDDIR),basename),DEFPIDEXT);
	} else {
	    pid_file = (char*) xmalloc(DEFPIDLEN+strlen(basename)+1);
	    pid_file = strcat(strcat(strcpy(pid_file,DEFPIDDIR),basename),DEFPIDEXT);
	}
    } else
	pid_forced = true;

    /* Check and verify the pid file */
    errno = 0;
    if (statn(pid_file, STATX_SIZE, &st) < 0) {
	if (errno != ENOENT) {
	    /* An other error like permission or HW problem */
	    warn("Can not stat %s: %s\n", pid_file, strerror(errno));

	} else if (pid_forced) {
	    /*
	     * Pid file was provided by the user therefore we assume that
	     * the proc is dead if the specified pid can not be veryfied.
	     */
	    if (remember_pids(pid_file,fullname,root,flags) < 0)
		exit(LSB_PROOFX);

	    if (list_empty(&remember))
	        exit(LSB_NOPROC);	/* New LSB: no pid file is no job */
	}
	free(pid_file);
	pid_file = NULL;

	/* No pid file means that we have to search in /proc/ */
    }

    if (pid_file && !st.st_size) {
	warn("Empty pid file %s for %s\n", pid_file, fullname);

	free(pid_file);
	pid_file = NULL;

	if (pid_forced)
	    exit(LSB_STATUS_NOPROC);

	/* No pid file means that we have to search in /proc/ */
    }

    if (pid_file) {		/* The case of having a pid file */
	if (verify_pidfile(pid_file,fullname,root,flags,false) < 0)
	    exit(LSB_PROOFX);
    } else {			/* No pid file found or given */
	if (pidof(fullname,root,flags) < 0)
	    exit(LSB_PROOFX);
    }

    /* Check and verify the ignore file */
    for (c = 0; (c < iargc) && (ignore_file = iargv[c]); c++) {
	errno = 0;
	if (statn(ignore_file, STATX_SIZE, &st) < 0) {
	    if (errno != ENOENT)
		warn("Can not stat %s: %s\n", ignore_file, strerror(errno));
	    continue;
	}
	if (!st.st_size) {
	    warn("Empty ignore file %s for %s\n", ignore_file, fullname);
	    continue;
	}
				/* The case of having a ignore file */
	if (verify_pidfile(ignore_file,fullname,root,flags,true) < 0)
	    exit(LSB_PROOFX);
    }

    if (iargc)
	clear_pids();		/* Remove all pids which should be ignored */

    /* Do main work */
    if (list_empty(&remember)) {
	/* killing a none existing process is already success.
	 * Nevertheless LSB says that we should note that if
	 * a signal is explicit given.
	 */
	exit(LSB_NOPROC);
    }

    num = 0;
    list_for_each(list, &remember) {
	PROC *proc = list_entry(list, PROC);
	do_kill(basename, proc->pid, snum, group_leader, process_group);
    }

    if (snum == SIGTERM || snum == SIGKILL) {
	int partsec = 5*wait;	/* We look 5 times within a second */
	/*
	 * Does anybody have a better idea ... something with sigaction()/signal()
	 * and alarm(). Who wakes us if the terminated process is finished?
	 * The process(es) is/are not a child of us.
	 */
	usleep(60*1000);	/* 60 ms time for the process and its childs */
again:
	if (check_pids(fullname,root,flags) < 0)
	    exit(LSB_PROOFX);

	if (list_empty(&remember))		/* success */
	    goto success;

	fflush(stdout);
	fflush(stderr);
	if (partsec-- > 0) {	/* sleep 0.2 seconds and try again */
	    usleep(2*100*1000);
	    goto again;
	}

	if (snum == SIGKILL)	/* SIGKILL was specified on the command line */
	    goto badterm;

	if (!sig_forced)	/* SIGTERM was specified on the command line */
	    goto badterm;

	if (check_pids(fullname,root,flags) < 0)
	    exit(LSB_PROOFX);

	list_for_each(list, &remember) {
	    PROC *proc = list_entry(list, PROC);
	    do_kill(basename, proc->pid, SIGKILL, group_leader, process_group);
	}

	/* Do we have killed them? */

	usleep(60*1000);	/* 60 ms time for the process and its childs */
	if (check_pids(fullname,root,flags) < 0)
	    exit(LSB_PROOFX);

	if (!list_empty(&remember))
	    goto badterm;

success:
	if (num) putchar('\n');
	errno = 0;
	if (pid_file && (unlink(pid_file) < 0)) {
	    if (errno != ENOENT)
		warn("Can not remove %s: %s\n", pid_file, strerror(errno));
	}
	exit(LSB_OK);

badterm:
	if (num) putchar('\n');
	exit(LSB_FAILED);
    }

    if (num)
	putchar('\n');
    exit(LSB_OK);

} /* end of main */

/* The core function */
static int do_kill(const char *inname, const pid_t proc, const int sig,
		   const int group_leader, const int process_group)
{
    pid_t target = proc;
    int stop = (sig == SIGTERM || sig == SIGKILL);

    errno = 0;
    if (group_leader) {
	if ((target = -getpgid(proc)) >= 0) {
	    if (errno != ESRCH)
		warn("Can not signal %s to process with pid %d: %s\n",
			signum_to_signame(sig), (int)proc, strerror(errno));
	    exit(LSB_PROOFX);
	}
    } else if (process_group)
	target = -proc;
#if DEBUG
    printf("kill(%d,%d)\n",(int)target, sig);
#else
#ifdef USE_BLOGD
    bootlog(B_NOTICE, "%s: kill(%d,%d)\n", we_are, (int)target, sig);
#endif
    if (stop) kill(target, SIGSTOP);
    errno = 0;
    if (kill(target, sig) < 0) {
	if (errno != ESRCH) {
	    warn("Can not signal %s to process with pid %d: %s\n",
		    signum_to_signame(sig), (int)proc,  strerror(errno));
	    exit(LSB_FAILED);
	}
    }
    if (stop) kill(target, SIGCONT);
    usleep(1);		/* Force the kernel to run the scheduler and update 
			   the environment of the current processes */
    if (!quiet) {
	if (num++) putchar(' ');
	printf("SIG%s %s(%d)",signum_to_signame(sig),inname,(int)proc);
    }
#endif
    return 0;
}
